src/IO/OpenedFileCache.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <map>
+#include <mutex>
+
+#include <Core/Types.h>
+#include <Common/ProfileEvents.h>
+#include <IO/OpenedFile.h>
+
+
+namespace ProfileEvents
+{
+    extern const Event OpenedFileCacheHits;
+    extern const Event OpenedFileCacheMisses;
+}
+
+namespace DB
+{
+
+
+/** Cache of opened files for reading.
+  * It allows to share file descriptors when doing reading with 'pread' syscalls on readonly files.
+  * Note: open/close of files is very cheap on Linux and we should not bother doing it 10 000 times a second.
+  * (This may not be the case on Windows with WSL. This is also not the case if strace is active. Neither when some eBPF is loaded).
+  * But sometimes we may end up opening one file multiple times, that increases chance exhausting opened files limit.
```

Comment:
No, I'm not sure.

It will help on some systems where increasing files limit is harder, e.g. developers' Mac OS machines.

Reply:
It will also help to make shared pool for asynchronous reads.

Reply:
I think it will be beneficial, see the example #25994.

Reply:
>It will help on some systems where increasing files limit is harder, e.g. developers' Mac OS machines.

Why it is harder there?

>I think it will be beneficial, see the example #25994.

Well, To me, complexity does not look like it worth it.

>It will also help to make shared pool for asynchronous reads.

Agree, although I would make it only for asynchronous reads then.

Reply:
I don't use Mac OS, but our docs contain somewhat complicated instruction:
https://clickhouse.tech/docs/en/development/build-osx/#caveats

I expect almost no one doing it.